packages/std/src/memory.rs

```diff
@@ -15,6 +15,22 @@ pub struct Region {
     pub length: u32,
 }
 
+impl Default for Box<Region> {
+    fn default() -> Self {
+        null_region()
+    }
+}
+
+impl Region {
+    pub fn as_ptr(&self) -> *const Region {
+        self as *const Region
+    }
```

Comment:
This API is problematic since a Region can freely move around in memory (what a `Box<Region>` does not do).

Reply:
I don't think I understand... `as_ptr()` works on a `self` reference, and so, that raw pointer is not going anywhere, except as a proxy or alias (a cast) to a reference to (it)`self` .

Reply:
> you mean is that we need a raw pointer to the reference to the `Box<Region>`, not to the `Region`?

Yes. Here is an example of the behaviour I tried to describe: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=98aae6f877da21e54ca3d9b5b8e3df92.

Region address 1: 0x7ffe614fae78
Region address 2: 0x7ffe614faed8
Boxed Region address 1: 0x5583b99a9c80
Boxed Region address 2: 0x5583b99a9c80

As you can see, the Region changes the address so we don't want `as_ref` to be available in Region instances just becuase we use it on `Box<Region>`.

Reply:
Yeah, you're right. What we need is to implement a `Trait` on top of `Box<Region>`. A trait that provides `to_ptr()` or similar.

Reply:
I would just keep them in the caller code to keep it explicit what is done

Reply:
I think those casts could be explicit, but are not clear.

Let's do as you say, but I would very much prefer a helper or member function
that does what needs to be done well than those.

Reply:
The more we discuss this, by the way, the more I like the initial `into_raw()` idea.
We're trying to deal with a raw pointer like if it wasn't one.

Reply:
The proper way to do this is with an `auto_ptr` type, that is released when it goes out of scope.

I did something similar for the jni-rs crate a while ago, by the way.

Reply:
> Let's do as you say, but I would very much prefer a helper or member function
> that does what needs to be done well, than those.

We can have a free helper function for the casts like:

```rust
/// Returns the address of the Region as an offset in linear memory
pub fn get_region_address(region: &Box<Region>) -> u32 {
    region.as_ref() as *const Region as u32
}
```

The argument type ensures the Box is only referenced and not consumed and dropped. This can then be used with:

```rust
        let iterator_id = unsafe {
            db_scan(
                start_region.as_ref().map(get_region_address).unwrap_or(0),
                end_region.as_ref().map(get_region_address).unwrap_or(0),
                order as i32,
            )
        };
```

>  We're trying to deal with a raw pointer like if it wasn't one.

We're trying to get a raw pointer that has a shorter lifetime than the object it points to. And I still don't think we should use manual creation and destruction for that. We just need to get scoping right.

> The proper way to do this is with an `auto_ptr` type, that is released when it goes out of scope.

Smart pointers probably won't work well over the import boundary where the pointers need to be casted to u32. At that point the address can be copied, such that the pointer can be destroyed before we're done using the object.

Reply:
> Smart pointers probably won't work well over the import boundary where the pointers need to be casted to u32. At that point the address can be copied, such that the pointer can be destroyed before we're done using the object.

That's something you can always do. That's a matter of usage, not functionality.

An `auto_ptr` will just guarantee the associated memory is freed, while allowing you to use `into_raw()` conveniently, that is, without worrying with manual destruction, and relying on scoping instead.

Reply:
> We just need to get scoping right.

Thinking about this a bit more, I think this is harder to do than I originally thought.

Coming from C++, scoping seems easy. Everything in a pair of `{`/`}` is a scope and objects are destructed in the reverse order of creation. You can move values out of a scope, but that is rare and explicit. In Rust in the other hand, you can easily write code where the objects are not destructed at the end of the scope but at any point in between. `Option::map` is such a case where the `Some` value is consumed, moved into the callback and destructed inside of the callback.

Reply:
Yes, exactly. That's why I like `into_raw()`. It's clearly saying: "Oh, you want a raw pointer, right? Well, the memory it points to is your problem now" (i.e. forget about scoping, the borrow checker, compilation checks, and whatnot).